File: _legacy/packages/cli/src/commands/sync.ts
import type { Command } from 'commander';
import chalk from 'chalk';
import ora from 'ora';
import { initDb, loadConfigAsync, ingestRepos, NotConfiguredError } from '@repog/core';

/**
 * Register the `repog sync` command.
 * Syncs repositories from GitHub to the local database.
 */
export function register(program: Command): void {
  program
    .command('sync')
    .description('Sync repositories from GitHub to local database')
    .option('--owned', 'Sync owned repositories')
    .option('--starred', 'Sync starred repositories')
    .option('--full-tree', 'Always fetch file trees regardless of README length')
    .option('--verbose', 'Show detailed progress per repository')
    .action(async (options) => {
      const { owned, starred, fullTree, verbose } = options;

      // Require at least one of --owned or --starred
      if (!owned && !starred) {
        console.error(chalk.red('Error: Specify --owned, --starred, or both.'));
        process.exit(1);
      }

      try {
        // Load config to get dbPath
        const config = await loadConfigAsync();
        if (!config.githubPat) {
          console.error(chalk.red('Run `repog init` first.'));
          process.exit(1);
        }

        // Initialize database
        initDb(config.dbPath);

        // Start the ingestion
        const spinner = verbose ? null : ora('Syncing repositories...').start();
        let syncedCount = 0;
        let skippedCount = 0;
        let errorCount = 0;

        for await (const event of ingestRepos({
          includeOwned: !!owned,
          includeStarred: !!starred,
          fullTree: !!fullTree,
          verbose: !!verbose,
        })) {
          switch (event.type) {
            case 'repo':
              syncedCount++;
              if (verbose) {
                const statusIcon = chalk.green('✓');
                const statusText = event.status === 'new' ? 'new    ' : 'updated';
                console.log(`${statusIcon} ${statusText}  ${event.repo}`);
              } else if (spinner) {
                spinner.text = `Syncing repositories... (${syncedCount} synced, ${skippedCount} skipped)`;
              }
              break;

            case 'skip':
              skippedCount++;
              if (verbose) {
                console.log(`${chalk.yellow('~')} skipped  ${event.repo}  (${event.reason})`);
              } else if (spinner) {
                spinner.text = `Syncing repositories... (${syncedCount} synced, ${skippedCount} skipped)`;
              }
              break;

            case 'error':
              errorCount++;
              if (verbose) {
                console.log(`${chalk.red('✗')} error    ${event.repo}     (${event.error})`);
              }
              break;

            case 'done':
              if (spinner) {
                spinner.stop();
              }
              if (event.errors > 0) {
                console.log(
                  chalk.yellow(
                    `✓ Sync complete — ${event.total} repos synced, ${event.skipped} skipped, ${event.errors} errors`
                  )
                );
              } else {
                console.log(
                  chalk.green(
                    `✓ Sync complete — ${event.total} repos synced, ${event.skipped} skipped, ${event.errors} errors`
                  )
                );
              }
              break;
          }
        }

        // Exit with code 1 if there were errors
        if (errorCount > 0) {
          process.exit(1);
        }
      } catch (error) {
        if (error instanceof NotConfiguredError) {
          console.error(chalk.red('Run `repog init` first.'));
          process.exit(1);
        }
        throw error;
      }
    });
}

File: commands/sync.go
package commands

import (
	"context"
	"fmt"
	"os"
	"time"

	"github.com/briandowns/spinner"
	"github.com/fatih/color"
	"github.com/spf13/cobra"

	"github.com/hackastak/repog/internal/config"
	"github.com/hackastak/repog/internal/db"
	"github.com/hackastak/repog/internal/provider"
	_ "github.com/hackastak/repog/internal/provider/gemini"
	_ "github.com/hackastak/repog/internal/provider/ollama"
	_ "github.com/hackastak/repog/internal/provider/openai"
	_ "github.com/hackastak/repog/internal/provider/openrouter"
	_ "github.com/hackastak/repog/internal/provider/voyageai"
	"github.com/hackastak/repog/internal/sync"
)

var syncCmd = &cobra.Command{
	Use:   "sync",
	Short: "Sync repositories from GitHub to local database",
	Long:  "Fetch repository metadata, READMEs, and file trees from GitHub and store them locally.",
	RunE:  runSync,
}

var (
	syncOwned    bool
	syncStarred  bool
	syncFullTree bool
	syncVerbose  bool
)

func init() {
	syncCmd.Flags().BoolVar(&syncOwned, "owned", false, "Sync owned repositories")
	syncCmd.Flags().BoolVar(&syncStarred, "starred", false, "Sync starred repositories")
	syncCmd.Flags().BoolVar(&syncFullTree, "full-tree", false, "Always fetch file trees regardless of README length")
	syncCmd.Flags().BoolVar(&syncVerbose, "verbose", false, "Show detailed progress per repository")
}

func runSync(cmd *cobra.Command, args []string) error {
	green := color.New(color.FgGreen).SprintFunc()
	yellow := color.New(color.FgYellow).SprintFunc()
	red := color.New(color.FgRed).SprintFunc()

	// Default to both if no flags specified
	if !syncOwned && !syncStarred {
		syncOwned = true
		syncStarred = true
	}

	// Load config
	cfg, err := config.LoadConfig()
	if err != nil {
		fmt.Fprintln(os.Stderr, red("Run `repog init` first."))
		os.Exit(1)
	}

	githubPAT, err := config.GetGitHubPAT()
	if err != nil {
		fmt.Fprintln(os.Stderr, red("Run `repog init` first."))
		os.Exit(1)
	}

	// Open database
	database, err := db.Open(cfg.DBPath, cfg.Embedding.Dimensions)
	if err != nil {
		fmt.Fprintln(os.Stderr, red("Database error:"), err)
		os.Exit(1)
	}
	defer func() { _ = database.Close() }()

	// Calculate chunk size based on embedding provider's token limit
	embeddingAPIKey, err := config.GetAPIKeyForProvider(cfg.Embedding.Provider)
	if err != nil {
		fmt.Fprintln(os.Stderr, red("Failed to get embedding API key:"), err)
		os.Exit(1)
	}

	embedProvider, err := provider.NewEmbeddingProvider(cfg.Embedding, embeddingAPIKey)
	if err != nil {
		fmt.Fprintln(os.Stderr, red("Failed to create embedding provider:"), err)
		os.Exit(1)
	}

	maxChunkSize := sync.CalculateMaxCharsFromTokens(embedProvider.MaxTokens())

	if syncVerbose {
		fmt.Printf("Using chunk size: %d characters (based on %d token limit)\n", maxChunkSize, embedProvider.MaxTokens())
	}

	// Start ingestion
	var s *spinner.Spinner
	if !syncVerbose {
		s = spinner.New(spinner.CharSets[14], 100*time.Millisecond)
		s.Suffix = " Syncing repositories..."
		s.Start()
	}

	var syncedCount, skippedCount, errorCount int

	eventCh := sync.IngestRepos(context.Background(), sync.IngestOptions{
		IncludeOwned:   syncOwned,
		IncludeStarred: syncStarred,
		FullTree:       syncFullTree,
		MaxChunkSize:   maxChunkSize,
		DB:             database,
		GitHubPAT:      githubPAT,
	})

	for event := range eventCh {
		switch event.Type {
		case "repo":
			syncedCount++
			if syncVerbose {
				statusText := "new    "
				if event.Status == "updated" {
					statusText = "updated"
				}
				fmt.Println(green("✓"), statusText, event.Repo)
			} else if s != nil {
				s.Suffix = fmt.Sprintf(" Syncing repositories... (%d synced, %d skipped)", syncedCount, skippedCount)
			}

		case "skip":
			skippedCount++
			if syncVerbose {
				fmt.Println(yellow("~"), "skipped ", event.Repo, "("+event.Reason+")")
			} else if s != nil {
				s.Suffix = fmt.Sprintf(" Syncing repositories... (%d synced, %d skipped)", syncedCount, skippedCount)
			}

		case "error":
			errorCount++
			if syncVerbose {
				fmt.Println(red("✗"), "error   ", event.Repo, "("+event.Reason+")")
			}

		case "done":
			if s != nil {
				s.Stop()
			}

			if event.Errors > 0 {
				fmt.Println(yellow("✓"), fmt.Sprintf("Sync complete — %d repos synced, %d skipped, %d errors",
					event.Total, event.Skipped, event.Errors))
			} else {
				fmt.Println(green("✓"), fmt.Sprintf("Sync complete — %d repos synced, %d skipped, %d errors",
					event.Total, event.Skipped, event.Errors))
			}
		}
	}

	if errorCount > 0 {
		os.Exit(1)
	}

	return nil
}
